src/Microsoft.ML.Data/Transforms/Hashing.cs

```diff
@@ -409,102 +409,91 @@ private ValueGetter<uint> ComposeGetterOne(Row input, int iinfo, int srcCol, Col
 
             if (srcType is KeyType)
             {
-                switch (srcType.RawKind)
-                {
-                    case DataKind.U1:
-                        return MakeScalarHashGetter<byte, HashKey1>(input, srcCol, seed, mask);
-                    case DataKind.U2:
-                        return MakeScalarHashGetter<ushort, HashKey2>(input, srcCol, seed, mask);
-                    case DataKind.U4:
-                        return MakeScalarHashGetter<uint, HashKey4>(input, srcCol, seed, mask);
-                    default:
-                        Host.Assert(srcType.RawKind == DataKind.U8);
-                        return MakeScalarHashGetter<ulong, HashKey8>(input, srcCol, seed, mask);
-                }
+                if (srcType.RawType == typeof(byte))
+                    return MakeScalarHashGetter<byte, HashKey1>(input, srcCol, seed, mask);
+                else if (srcType.RawType == typeof(ushort))
+                    return MakeScalarHashGetter<ushort, HashKey2>(input, srcCol, seed, mask);
+                else if (srcType.RawType == typeof(uint))
+                    return MakeScalarHashGetter<uint, HashKey4>(input, srcCol, seed, mask);
+
+                Host.Assert(srcType.RawType == typeof(ulong));
+                return MakeScalarHashGetter<ulong, HashKey8>(input, srcCol, seed, mask);
             }
 
-            switch (srcType.RawKind)
-            {
-                case DataKind.U1:
-                    return MakeScalarHashGetter<byte, HashU1>(input, srcCol, seed, mask);
-                case DataKind.U2:
-                    return MakeScalarHashGetter<ushort, HashU2>(input, srcCol, seed, mask);
-                case DataKind.U4:
-                    return MakeScalarHashGetter<uint, HashU4>(input, srcCol, seed, mask);
-                case DataKind.U8:
-                    return MakeScalarHashGetter<ulong, HashU8>(input, srcCol, seed, mask);
-                case DataKind.U16:
+            if (srcType.RawType == typeof(byte))
```

Comment:
>if (srcType.RawType == typeof(byte)) [](start = 12, length = 36)

It used to not matter as much since it was a case statement, but now that it's a chained if statement we probably should give some thought to what is the most common case. The most common case for hashing is text, so we should probably check that *first*, rather than letting it fall through as the absolute last case when all else fails.

Also probably this would probably be a good time to to checks against the `ColumnType` instances themselves -- the original author did things in fashion because they considered it to be more concise, but I'm not sure that reasoning holds any longer.

Reply:
> The most common case for hashing is text, so we should probably check that first, 

Good call. Fixing.

> Also probably this would probably be a good time to to checks against the ColumnType instances themselves

In my mind, there are 2 reasons why I like comparing the types here, and not against the ColumnType instances.

1. It is nice and easy to match up the `if (.... == typeof(byte))` with the next line `MakeScalarHashGetter<byte,....`. You can match the `byte` and `byte` and easily know you are using the same type.
2. It is reflective of the code above for `KeyType`, where you can't actually check for `NumberType.U4`, since it is a `KeyType`.

I don't feel strongly about these reasons, but they do seem to be nice. If you'd like it the other way, I can change it. Just let me know if you feel strongly for it to be changed.

---
In reply to: [248925374](https://github.com/dotnet/machinelearning/pull/2176#discussion_r248925374) [](ancestors = 248925374)

Reply:
Fair enough, I think your reasoning works.

---
In reply to: [249119279](https://github.com/dotnet/machinelearning/pull/2176#discussion_r249119279) [](ancestors = 249119279,248925374)